src/vm/database/sqlite.rs

```diff
@@ -32,30 +32,92 @@ fn sqlite_has_entry(conn: &Connection, key: &str) -> bool {
     sqlite_get(conn, key).is_some()
 }
 
-impl KeyValueStorage for SqliteConnection {
-    fn put(&mut self, key: &str, value: &str) {
+impl SqliteConnection {
+    pub fn put(&mut self, key: &str, value: &str) {
         sqlite_put(&self.conn, key, value)
     }
 
-    fn get(&mut self, key: &str) -> Option<String> {
+    pub fn get(&mut self, key: &str) -> Option<String> {
         sqlite_get(&self.conn, key)
     }
 
-    fn has_entry(&mut self, key: &str) -> bool {
+    pub fn insert_metadata(&mut self, bhh: &BlockHeaderHash, contract_hash: &str, key: &str, value: &str) -> bool {
+        debug!("insert_metadata: {}, {}, {}", bhh, contract_hash, key);
+
+        let blockhash = bhh.to_hex();
+        let key = format!("clr-meta::{}::{}", contract_hash, key);
+        let params: [&dyn ToSql; 3] = [&blockhash, &key, &value.to_string()];
+
+        match self.conn.execute("INSERT INTO metadata_table (blockhash, key, value) VALUES (?, ?, ?)", &params) {
+            Ok(_) => true,
+            Err(SqliteError::SqliteFailure(e, _)) => {
+                if e.code == SqliteErrorCode::ConstraintViolation {
```

Comment:
Is this code path allowed when a peer is processing blocks?  How could this happen?

Reply:
Non-unique `(blockhash, key)`

Reply:
Addressed in 505a1cd -- I changed it to panic on all SqliteErrors, an attempt to write metadata that has already been written should at this point panic.

Reply:
I'm pretty sure there are legitimate ways to arrive at a non-unique `(blockhash, key)`.  Consider two contracts with identical code that get instantiated in the same block, and consider two contract-calls on them that each set the same variable or the same data map.  Won't this cause an undesired panic?

Reply:
No -- metadata is keyed by `blockhash` and then the metadata `key`, which is actually `contract-identifier, metadata-key`. Contract identifiers (which are publisher address, name) are unique per-fork.

Reply:
Ah, I didn't make that connection.  This is fine by me then :+1: